### PR TITLE
Trb 44 without tests

### PR DESCRIPTION
**Pull request by pitner33**
WARNING! Mindfck inside! :)

### CODE REVIEW

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        } else if (isItBuildingToUpgrade(allProgressionModelsList.get(i))) {
+          upgradeBuildingFromProgression(allProgressionModelsList.get(i));
+        } else if (isItTroopToUpgrade(allProgressionModelsList.get(i))) {
+          upgradeTroopFromProgression(allProgressionModelsList.get(i));
+        }
+        safeDeleteProgression(allProgressionModelsList.get(i));
+      }
+    }
+  }
+
```
Comment (lyancsie) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:78]: //TODO: Refactor in the future

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      addBuildingToKingdom(progression, newBuilding);
+    } else if (type.equals("farm")) {
+      Building newBuilding = BuildingFactory.createBuilding(BuildingType.farm);
+      addBuildingToKingdom(progression, newBuilding);
+    } else if (type.equals("mine")) {
+      Building newBuilding = BuildingFactory.createBuilding(BuildingType.mine);
+      addBuildingToKingdom(progression, newBuilding);
+    } else throw new BuildingTypeNotValidException("There is no such type of building");
+  }
+
```
Comment (lyancsie) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:118]: //TODO: refactor in the future

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    String buildingType = progression.getType();
+    if (buildingType.equals("barracks")) {
+      buildingService.upgradeBarracks(buildingToUpgrade);
+    } else if (buildingType.equals("farm")) {
+      buildingService.upgradeFarm(buildingToUpgrade);
+    } else if (buildingType.equals("mine")) {
+      buildingService.upgradeMine(buildingToUpgrade);
+    } else throw new NotValidTypeException("Invalid Building type");
+     //TODO TRB-29, 49, 50 for buildingUpgrade
+  }
```
Comment (lyancsie) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:138]: //TODO: refactor in the future

File: src/main/java/com/greenfox/tribes1/Troop/TroopService.java [hunk 4]
```diff
@@ -26,4 +30,14 @@ public boolean isValidTroop(Troop troop) {
     return troop != null;
   }
 
+  public Troop findById(Long id) throws TroopIdNotFoundException {
+    return Optional.of(troopRepository.findById(id)).get().orElseThrow(()
+            -> new TroopIdNotFoundException(("There is no Troop with such Id")));
+  }
```
Comment (lyancsie) [inline src/main/java/com/greenfox/tribes1/Troop/TroopService.java:36]: 💯 🥇 👍

File: src/main/resources/db/migration/mysql/V1_5__init.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  `is_create`   boolean,
+  `type` varchar (20),
+  `finished_at` timestamp(6),
+
+  PRIMARY KEY (`id`)
+);
+
+CREATE TABLE `kingdom_progression` (
+  `kingdom_id`          bigint(20) not null,
+  `progression_id` bigint(20) NOT NULL,
```
Comment (lyancsie) [inline src/main/resources/db/migration/mysql/V1_5__init.sql:15]: unnecessary NOT NULL at progression_id

Comment (lyancsie) [reply src/main/resources/db/migration/mysql/V1_5__init.sql]: Is it going to be @ManyToMany?
There is no PRIMARY KEY in the joined table

File: src/main/resources/db/migration/mysql/V1_5__init.sql [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  PRIMARY KEY (`progression_id`),
+  KEY progression_id (`progression_id`),
```
Comment (lyancsie) [inline src/main/resources/db/migration/mysql/V1_5__init.sql:17]: I'm not sure if this is needed

**Review by refike — APPROVED** [review]

**Review by lyancsie — APPROVED** [review]
Discussed everything.

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+  public void safeDeleteProgression (Progression progression) {
+    progression.setKingdom(null);
+    progressionRepository.delete(progression);
+  }
+
+  public Progression findById (Long id) throws ProgressionIdNotFoundException {
+    return Optional.of(progressionRepository.findById(id)).get().orElseThrow(()
+            -> new ProgressionIdNotFoundException(("There is no Troop with such Id")));
```
Comment (refike) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:53]: Probably it's going to respond with Building in the message as well.

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    return (progression.getType().equals("barracks") ||
+            progression.getType().equals("farm") ||
+            progression.getType().equals("mine"));
+  }
+
+  public Boolean isTypeTroop (Progression progression) {
+    return progression.getType().equals("troop");
+  }
+
+  public void createNewBuildingFromProgression_AndAddItToKingdom (Progression progression) throws BuildingTypeNotValidException {
```
Comment (ChrisFrill) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:105]: Probably need a nice refactor in the future to apply SRP to our code.

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    return Optional.of(progressionRepository.findById(id)).get().orElseThrow(()
+            -> new ProgressionIdNotFoundException(("There is no Troop with such Id")));
+  }
+
+  public List<Progression> findAllProgressionModel() {
+    return progressionRepository.findAll();
+  }
+
+//TODO Progression SQL!!!!!!
+  public void findProgressionsWithExpiredTimestamp_CreateOrUpgradeModelFromThem_DeleteThem() throws BuildingTypeNotValidException, TroopTypeNotValidException, TroopIdNotFoundException, BuildingNotValidException, NotValidTypeException, BuildingIdNotFoundException, TroopNotValidException {
```
Comment (ChrisFrill) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:61]: Probably need a nice refactor in the future to apply SRP to our code.

File: src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    } else if (type.equals("farm")) {
+      Building newBuilding = BuildingFactory.createBuilding(BuildingType.farm);
+      addBuildingToKingdom(progression, newBuilding);
+    } else if (type.equals("mine")) {
+      Building newBuilding = BuildingFactory.createBuilding(BuildingType.mine);
+      addBuildingToKingdom(progression, newBuilding);
+    } else throw new BuildingTypeNotValidException("There is no such type of building");
+  }
+
+  public void createNewTroopFromProgression_AndAddItToKingdom (Progression progression) throws BuildingTypeNotValidException, TroopTypeNotValidException {
```
Comment (ChrisFrill) [inline src/main/java/com/greenfox/tribes1/Progression/ProgressionService.java:119]: Probably need a nice refactor in the future to apply SRP to our code.

**Review by ChrisFrill — APPROVED** [review]

**Review by csadesz — APPROVED** [review]

**Pull request merged**